### PR TITLE
storage/etcd: skip SendInitialEvents if the request is backward compatible

### PR DESCRIPTION
**Pull request by p0lyn0mial**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
It turns out that the watch-cache is disabled by default for `events` resource.

When the WatchList feature is enabled and a watch request for events with empty ListOptions is send. Then default options are applied and the request is passed to the etcd implementation. In the end the request fails because etcd doesn't implement streaming API for alpha. 

This PR provides a fix that allows for skipping SendInitialEvents when the request is considered to be backward compatible (thanks @wojtek-t for suggesting it!).

Backward compatibility is defined as RV = "" || RV = "O" and AllowWatchBookmark is set to false. 
In that case we rely on https://github.com/kubernetes/kubernetes/blob/267eb25e60955fe8e438c6311412e7cf7d028acb/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go#L260

#### Which issue(s) this PR fixes:

Fixes #https://github.com/kubernetes/kubernetes/issues/116357

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher_test.go [hunk 1]
```diff
@@ -99,6 +102,15 @@ func TestProgressNotify(t *testing.T) {
 	storagetesting.RunOptionalTestProgressNotify(ctx, t, store)
 }
 
+func TestSendInitialEventsBackwardCompatibility(t *testing.T) {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher_test.go:105]: The idea is that we're trying as many tests implementation-agnostic as possible and run the exact same tests against both etcd3 and watchcache.

It seems to be one of those tests, so let's implement it in pkg/storage/testing/
and run it for both etcd3 and watchcache (pkg/storage/tests/cacher_test.go).

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 2]
```diff
@@ -865,8 +865,12 @@ func growSlice(v reflect.Value, maxCapacity int, sizes ...int) {
 }
 
 // Watch implements storage.Interface.Watch.
+// TODO(#115478): for beta the WatchList feature also must/should support streaming.
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:868]: This TODO is somewhat misleading to me. Watch by definition supports streaming :)

We probably want to say sth like:
TODO(...): For beta of WatchList feature, etcd3 implemenation should always support it.

[(or sth like that)]

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher_test.go (hunk 1 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher_test.go:105]: Yes - see above, all the tests above are following this pattern.

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher_test.go:102]: okay, I think I have it. PTAL

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go (hunk 2 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: done, reworded the comment.

**Comment by p0lyn0mial** [comment]
> I'm holiding this until we have alpha suite passing, feel free to cancel once it passes.

`pull-kubernetes-e2e-gce-cos-alpha-features` is green, releasing the hold.

/hold cancel

**Comment by p0lyn0mial** [comment]
xref: https://github.com/kubernetes/enhancements/issues/3157

**Review by hashim21223445 — APPROVED** [review]

**Pull request merged**